python/ray/serve/controller.py

```diff
@@ -439,15 +446,78 @@ def get_root_url(self):
                 )
         return http_config.root_url
 
+    def config_checkpoint_deleted(self) -> bool:
+        """Returns whether the config checkpoint has been deleted.
+
+        Get the config checkpoint from the kv store. If it is None, then it has been
+        deleted.
+        """
+        return self.kv_store.get(CONFIG_CHECKPOINT_KEY) is None
+
     def shutdown(self):
-        """Shuts down the serve instance completely."""
+        """Shuts down the serve instance completely.
+
+        This method will only be triggered when `self._shutting_down` is true. It
```

Comment:
Should we retry the shutdown if any resources fail to be deleted? Currently, the behavior is that if any resource fails to delete, the controller won't be deleted.

Reply:
Yea, this `shutdown()` is called in the `run_control_loop()` now and will only kill the controller if everything else are shut down 🙂